---
title: "Installing and Using BrainSABER"
author: "Carrie Minette"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
 %\VignetteIndexEntry{BrainSABER} 
  %\VignetteEncoding{UTF-8} 
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

The Allen Institute for Brain Science provides an 
RNA sequencing (RNA-Seq) data resource for studying 
transcriptional mechanisms involved in human brain 
development known as BrainSpan. This resource serves as an 
additional control in research involving RNA sequencing of 
human brain tissue. BrainSABER facilitates comparisons of user data
with the various developmental stages and brain structures found in
the BrainSpan atlas through an R Shiny interface. BrainSABER utilizes the Bioconductor
ExpressionSet class as a self-validating container for user data
and produces similarity matrices containing samples of the userâ€™s data
analyzed against each sample in the BrainSpan database. These matrices 
are presented as dynamic heat maps, and include both cosine and Euclidean 
similarity methods.

# Installing BrainSABER

BrainSABER requires Bioconductor 3.11 or higher, and the AIBSARNA data package, which contains the Allen Institute's Brain Atlas gene expression data gathered into a Biobase ExpressionSet.  For more information on Bioconductor, please see their website at https://bioconductor.org.  To install Bioconductor, AIBSARNA, and BrainSABER, run the following commands in R or RStudio:

```{r, eval=FALSE}
# for R version >= 3.5
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("AIBSARNA")
BiocManager::install("BrainSABER")
```


After installing, attach the BrainSABER package with the following command:

```{r loadpkg, message=FALSE, warning=FALSE, results='hide'}
library(BrainSABER)
```


# Preparing an ExpressionSet object

The BrainSABER workflow is based around an instance of Biobase ExpressionSet class, which provides a common interface familiar to those who have analyzed microarray experiments with Bioconductor. The class requires the following three input files:

1. AssayData, a numeric matrix of expression values, where rows are genes, and columns are cells/samples

2. phenoData, an AnnotatedDataFrame object, where rows are cells/samples, and columns are cell/sample attributes (such as cell type, culture condition, day captured, etc.)

3. featureData, an AnnotatedDataFrame object, where rows are features (e.g. genes), and columns are gene attributes, such as gene identifiers, gc content, etc.

The expression value matrix must have the same number of columns as the phenoData has rows, and it must have the same number of rows as the featureData data frame has rows. Row names of the phenoData object should match the column names of the expression matrix. Row names of the featureData object should match row names of the expression matrix.  

## Loading the Data

For this vignette, we will construct a toy ExpressionSet object by extracting 50 samples and 200 genes from AIBSARNA, using the code below:

```{r getSampleData, message=FALSE, warning=FALSE, results='hide'}
# load AIBSARNA data package 
library(AIBSARNA)
data("AIBSARNA")

# Obtain the sample indexes to use for subsetting (not random)
sample_idx <- 1:50 * 10 - 1
# Set the RNG seed for repeatable results
set.seed(8)
# Get the total number of genes available
totalGenes <- nrow(AIBSARNA)
# Sample the indices of 200 random genes
gene_idx <- sample.int(totalGenes, 200)

# Subset AIBSARNA
toy_exprs <- exprs(AIBSARNA)[gene_idx, sample_idx]
toy_fd <- fData(AIBSARNA)[gene_idx, ]
toy_pd <- pData(AIBSARNA)[sample_idx, ]

# Convert feature and pheno data to AnnotatedDataFrame required by ExpressionSet

toy_featureData <- AnnotatedDataFrame(data = toy_fd)
toy_phenoData <- AnnotatedDataFrame(data = toy_pd)

# Create toy ExpressionSet
toySet <- ExpressionSet(assayData = toy_exprs,
                        featureData = toy_featureData,
                        phenoData = toy_phenoData)

```


# The BrainSABER Workflow (Shiny app)

BrainSABER comes with a Shiny app that runs the entire BrainSABER workflow, and also generates heatmaps based on the similarity matrices produced by BrainSABER.  This Shiny app requires three .CSV files containing the data required to construct an ExpressionSet.  We can save the toy data we subsetted earlier as .CSV files to use with the Shiny app.  The following code will save these files in your current working directory.

```{r saveCSV}
write.csv(toy_exprs, file = "toy_expression_matrix.csv")
write.csv(as.data.frame(toy_fd), file = "toy_gene_identification_data.csv")
write.csv(as.data.frame(toy_pd), file = "toy_sample_information_data.csv")
```


Once these files are saved, we can run the Shiny app.  The app will guide us through the BrainSABER workflow.
```{r runShinyBrainsABER, eval=FALSE}
runShinyBrainSABER()
```

## Session Info
```{r sessionInfo}
sessionInfo()
```